### ISSUE TITLE
[swift] Generated code crashes if data does not match spec

### ISSUE THREAD

**Issue by philippelatulippe** [swagger-api/swagger-codegen#4382]
##### Description

Any error in the data being decoded (like incorrect type, missing key, parsing error in the date, etc.) will cause the application to crash. The decoders (in *Models.swift*) make heavy use of fatalError() and force-unwrapping. Is there a reason for this? Is there a decoder, or decoder option, that supports graceful deserialization error handling? 

Is there interest in a patch that would fix this? I was thinking of having all decode functions return an enum with "success" and "failure" cases, something like:

````
enum Decoded<T> {
    case success(T)
    case failure(DecoderError)
}
````

##### Swagger-codegen version

master

**Comment by wing328** [comment]
@philippelatulippe can you share a spec to reproduce the issue? I'm assuming you're using the latest master with this issue.

**Comment by philippelatulippe** [comment]
I'm using the petstore spec and the latest master, but any spec will do. Just feed the decoder any JSON data that is not expected by the spec and it will call `fatalError()` and terminate the application.  

Here's a quick way to test it with the PetStoreClient. The spec says that "name" must be a string, but here the value in the json is an integer. 

        let jsonData = "{\"name\": 5}".data(using: .utf8)!
        let json = try! JSONSerialization.jsonObject(with: jsonData, options: .mutableContainers)
        var pet: Pet = Decoders.decode(clazz: Pet.self, source: json as AnyObject)

The application then crashes in the function `Decoders.decode<T>(clazz: T.Type, source: AnyObject) -> T`, [which calls fatalError](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/swift3/Models.mustache#L92) if the type doesn't match the spec. This isn't ideal in production, since it's not so rare for servers to accidentally send data that deviates from the spec. We don't want the application to crash hard if it receives the wrong data.

The generated code doesn't seem to have a way to deal with that. I'm proposing having all the decode() functions return an enum with two cases: success, which will contain a result, and failure, which will contain an error. Callers can then use a switch to handle the two cases. For optional values, a convenience function would return the value directly, and nil if there is an error.

The main `decode` function liked above would then look something like:

    static func decode<T>(clazz: T.Type, source: AnyObject) -> Decoded<T> {
        initialize()
        if let value = source.int32Value as? T, source is NSNumber, T.self is Int32.Type {
            return .success(value)
        }
        if let value = source.int32Value as? T, source is NSNumber, T.self is Int64.Type {
            return .success(value)
        }
        if let intermediate = source as? String, let value = UUID(uuidString: intermediate) as? T, source is String, T.self is UUID.Type {
            return .success(value)
        }
        if let value = source as? T {
            return .success(value)
        }
        //The last two expressions in this condition must be unnecessary, but it would be good to test this.
        if let intermediate = source as? String, let value = Data(base64Encoded: intermediate) as? T, T.self is Data.Type, source is String {
            return .success(value)
        }

        let key = "\(T.self)"
        if let decoder = decoders[key] {
           return decoder(source)
        } else {
            return .failure(.typeMismatch(expected: String(describing: clazz), actual: String(describing: source)))
        }
    }

Is this approach okay for swagger-codegen? This is loosely based off [Argo](https://github.com/thoughtbot/Argo/), without the functional programming. Another way to do it would be to have the decode functions throw when they hit an error.

**Comment by jaz-ah** [comment]
seems very reasonable.

**Comment by wing328** [comment]
Closed via #4547
